[mlir][vector] Add vector.to_elements unrolling

amd-eochoalo opened the pull request:
This PR adds support for unrolling `vector.to_element`'s source operand.

It transforms

```mlir
%0:8 = vector.to_elements %v : vector<2x2x2xf32>
```

to

```mlir
%v0 = vector.extract %v[0] : vector<2x2xf32> from vector<2x2x2xf32>
%v1 = vector.extract %v[1] : vector<2x2xf32> from vector<2x2x2xf32>
%0:4 = vector.to_elements %v0 : vector<2x2xf32>
%1:4 = vector.to_elements %v1 : vector<2x2xf32>
// %0:8 = %0:4 - %1:4
```

This pattern will be applied until there are only 1-D vectors left.

hanhanW committed:
[mlir][vector] Add support for lowering n-D vector.to_elements op.

The revision adds a pattern that flattens 2 or more dimensional
`vector.to_elements` ops by `vector.shape_cast` + `vector.to_elements`.

It also adds the lowering pattern to ConvertVectorToLLVMPass and
complete the tests.

It recovers the e2e lowering breakage from https://github.com/llvm/llvm-project/commit/b4c31dc98dfc929728904cd96f0f4cf812c4d5b5 on LLVM path.

Signed-off-by: hanhanW <hanhan0912@gmail.com>

amd-eochoalo committed:
[mlir][vector] Add function to unroll vectors.

Extract n vector<axbx...> from vector<nxaxbx...>

This patch adds a utility function that will unroll vector values.
This is different from the current utility function that focuses
on unrolling vector operations.

amd-eochoalo committed:
[mlir][vector] Add vector.to_elements unrolling.

amd-eochoalo committed:
Add new populate patterns for flattening.

amd-eochoalo committed:
Split tests for unrolling and flattening to elements

amd-eochoalo committed:
Fix test

amd-eochoalo committed:
Address review comments

amd-eochoalo committed:
Remove unnecessary comment

amd-eochoalo commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorToElements.cpp:
```diff
@@ -0,0 +1,81 @@
+
+  LogicalResult matchAndRewrite(vector::ToElementsOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value> vectors;
```
It's first use is the line below. It is passed in as a reference to the function being called.

amd-eochoalo committed:
Address review comments

amd-eochoalo commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorToElements.cpp:
```diff
@@ -0,0 +1,81 @@
+    if (LogicalResult match =
+            vector::unrollVectorValue(op.getSource(), rewriter, vectors);
+        failed(match)) {
+      return match;
```
Thanks! This made me look into the definition of `notifyMatchFailure` and I see that it the reason is a side-effect and not part of the returned value. Although, it does feel a bit weird that the error message is separate from the value returned here. Like, we can set the error message and then continue along. https://github.com/llvm/llvm-project/pull/157142/commits/244eed587f232ae8c67af133887b6e6f3e8e351d

kuhar reviewed (approved):
LGTM but let's wait for a at least one more review

newling reviewed (changes requested):
Can you please make separate PRs for unrolling and flattening?

amd-eochoalo committed:
Disable and move vector flattening

amd-eochoalo committed:
Re-enable and rename flattening to linearize

amd-eochoalo committed:
Remove LinearizeToElements

amd-eochoalo committed:
Add comment

newling commented on mlir/include/mlir/Dialect/Vector/Utils/VectorUtils.h:
```diff
@@ -255,6 +255,16 @@ using UnrollVectorOpFn =
+/// 2. Check for scalable dimensions. If so, return failure.
+/// 3. Returns the values of n vector.extract operations corresponding
+///    to the outermost dimension.
+LogicalResult unrollVectorValue(Value vector, PatternRewriter &rewriter,
```
In terms of the most likely to be reused utility API, could consider 

```c++
/// Generic utility for unrolling values of type vector<NxAxBx...>
/// to N values of type vector<AxBx...> using vector.extract. If the input
/// is rank-0 or has leading scalable dimension, failure is returned.
FailureOr<SmallVector<Value>> unrollVectorValue(TypedValue<VectorType>, RewriterBase &);
```

main differences being 
- unroll rank-1 to rank-0. Unless there is a reason for the generic case to stop at rank-1? 
- Make the constraint that it's a vector explicit with TypedValue
- Use the most general builder possible (this requires dropping the notifyMatchFailure though, so might not be ideal)


Also, consider using an InsertionGuard inside the function.

amd-eochoalo replied:
Hi @newling:

* I don't think we want to unroll rank-1 to rank-0. To answer your high level question, if I undestand correctly, this pattern is supposed to mimic the lowering strategy for `vector.from_elements` (seen here https://github.com/llvm/llvm-project/pull/151175 ) but with `vector.to_elements`. According to @Groverkss we should be using unrolling as opposed to shape casting as the default option. And I think the motivation is indeed for supporting lowering rank-2+ for different targets (i.e., SPIR-V and LLVM).
* Regarding using an InsertionGuard inside the function, I would agree if there was an operation. But for values, these might be parameters being passed into a block or a function. If the value being passed in does not have a definition, do you want me to place the operations at the beginning of the current block? I think not using an InsertionGuard in this case makes sense, but happy to make changes.

newling replied:
> These guards are helpful when the insertion point is modified (to make sure that it gets restored auto-magically). But in this PR, we don't modify it, do we?

I guess not, no. Rereading Erick's comment again I see now that the current design is good.

amd-eochoalo replied:
@banach-space 

> Where is the default encoded?

Maybe we should let @Groverkss comment himself. I am paraphrasing him from a comment he made on discord and possibly another one here on github. This "default" may be to IREE and not MLIR itself. I also believe that @Groverkss wants both of these patterns upstream for backends to choose. See this [comment](https://github.com/llvm/llvm-project/pull/151175#issuecomment-3168195047)

> How we ideally want to structure conversion to backends is:
>
> * Set of patterns to do unrolling from N-D vectors to 1-D vectors
> * Set of patterns to do flattening from N-D vectors to 1-D vectors (in case someone wants to do this, we dont have patterns for this today)
> * [continues]

newling reviewed (commented):
High-level question: Is the motivation for this PR that without it we currently fail to lower to rank-2+ from_elements ops from Vector to LLVM?

amd-eochoalo committed:
Improve API of unrollVectorValue.

Parameters are now:
* using TypedValue<VectorType> instead of just Value
* using RewriterBase class.

Return types are:
* changed to FailureOr<SmallValue<Value>> instead of passing
  a Value as a parameter and returning Logical.

banach-space reviewed (commented):
Thanks, just some minor comments.

amd-eochoalo commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -808,6 +808,28 @@ struct TestUnrollVectorFromElements
   }
 };
 
+struct TestUnrollVectorToElements
+    : public PassWrapper<TestUnrollVectorToElements,
+                         OperationPass<func::FuncOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestUnrollVectorToElements)
+
+  StringRef getArgument() const final {
+    return "test-unroll-vector-to-elements";
+  }
+  StringRef getDescription() const final {
+    return "Test unrolling patterns for to_elements ops";
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<func::FuncDialect, vector::VectorDialect>();
+  }
+
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    populateVectorToElementsLoweringPatterns(patterns);
+    (void)applyPatternsGreedily(getOperation(), std::move(patterns));
+  }
+};
+
```
The reason why this test pass was added is to keep the style consistent with https://github.com/llvm/llvm-project/pull/151175 . Sorry, I am not familiar with TD as an acronym. Do you mean the test dialect or the transform dialect? I am assuming you are asking about adding a transform dialect operation for this pass? I have no experience with them and how scaling is different between the transform dialect and these test passes here. Happy to add it if needed, but I prefer to maintain the consistency with the vector.from_elements unrolling PR.

banach-space replied:
>  Sorry, I am not familiar with TD as an acronym. 

Transform Dialect :)

> The reason why this test pass was added is to keep the style consistent with https://github.com/llvm/llvm-project/pull/151175 .

Consistency is important 👍🏻 However, note that that PR also added `apply_patterns.vector.unroll_from_elements` TD Op. IMO, we should remove the pass added in https://github.com/llvm/llvm-project/pull/151175 and instead have one TD Op for unrolling `to_elements` and `from_elements`.

>  Happy to add it if needed, but I prefer to maintain the consistency with the vector.from_elements unrolling PR.

Yes, consistency is key. Again, my suggestion is to only have a TD Op for both Ops.

amd-eochoalo commented on mlir/test/Dialect/Vector/vector-to-elements-lowering.mlir:
Sure, I can change it. Again, most of this is to keep the style consistent with vector.from_elements lowering which adds a vector-from-elements-lowering.mlir file.

newling replied:
My preference would be to keep the file name here with 'lowering' in it. In my mind, it is just an implementation decision that lowering is implemented with unrolling and not something else here (linearizing). 

What is potentially confusing (certainly tripped me up!) is that there is unrolling implemented for all ops [here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp) that unrolls with insert_strided_slice and extract_strided_slice (and preserves ranks). And then there is alot of unrolling with vector.extract and vector.insert that happens in the LowerVector*cpp files.

banach-space replied:
> My preference would be to keep the file name here with 'lowering' in it. In my mind, it is just an implementation decision that lowering is implemented with unrolling and not something else here (linearizing).

What I find confusing is that "_lowering" implies that "unrolling" =  _the lowering_, while it's just one possible/supported approach. Feel free to keep it as is if I'm the only one that's confused.

> What is potentially confusing (certainly tripped me up!) is that there is unrolling implemented for all ops [here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp) that unrolls with insert_strided_slice and extract_strided_slice (and preserves ranks). And then there is alot of unrolling with vector.extract and vector.insert that happens in the LowerVector*cpp files.

Not sure about the `extract` vs `extract_strided_slice` part (thanks for pointing out!), but I would start by using consistent naming for everything that's intended as unrolling :)

amd-eochoalo committed:
Documentation

amd-eochoalo committed:
Add transform.apply_patterns.vector.unroll_to_elements

amd-eochoalo committed:
Minor changes

banach-space reviewed (approved):
My final asks are minor, so approving as it, but please address them before landing.

Thanks for working on this! 🙏🏻 

LGTM % outstanding asks

amd-eochoalo committed:
Remove comment and use inline storage for SmallVector

amd-eochoalo committed:
Add test with transform interpreter

amd-eochoalo committed:
Use transform dialect library file

amd-eochoalo committed:
Merge branch 'main' into eochoa/2025-09-05/vector-to-elements-lowering

amd-eochoalo committed:
Update mlir/test/Dialect/Vector/lit.local.cfg

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>
